semgrep/semgrep/output.py

```diff
@@ -173,3 +185,167 @@ def build_output(
     else:
         # https://github.com/python/mypy/issues/6366
         raise RuntimeError(f"Unhandled output format: {type(output_format).__name__}")
+
+
+class OutputSettings(NamedTuple):
+    output_format: OutputFormat
+    output_destination: Optional[str]
+    quiet: bool
+    error_on_findings: bool
+
+
+@contextlib.contextmanager
+def managed_output(output_settings: OutputSettings) -> Generator:  # type: ignore
+    """
+    Context manager to capture uncaught exceptions &
+    """
+    output_handler = OutputHandler(output_settings)
+    try:
+        yield output_handler
+    except Exception as ex:
```

Comment:
I think we can avoid the indirection with `final_error` by doing something like:

```python
    try:
        yield output_handler
    except SemgrepError as ex:
        output_handler.close()
        print_error(str(ex))
        raise
    except Exception:
        output_handler.close()
        raise
    else:
        output_handler.close()
```

A bit wonky with `output_handler.close()` in each block, but it's necessary for the different print ordering with `SemgrepError`.

Reply:
I want to eventually be able to get `final_error` into the JSON output so passing it in was intentional.

Reply:
Ahh, that makes sense :+1: